File: HelloARKit/HelloARKit/ViewController.m
//
//  ViewController.m
//  HelloARKit
//
//  Created by zhoufei on 2023/8/4.
//

#import "ViewController.h"
#import "ARMoveSCNViewController.h"
#import "ARRotationSCNViewController.h"
#import "ARPlanSCNViewController.h"

@interface ViewController ()
- (IBAction)arButtionAction:(UIButton *)sender;
@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
}

- (IBAction)arButtionAction:(UIButton *)sender {
     
    switch (sender.tag) {
        case 10:
        {
            ARPlanSCNViewController *vc = [ARPlanSCNViewController new];
            [self presentViewController:vc animated:YES completion:nil];
        }
            break;
        case 20:
        {
            ARMoveSCNViewController *vc = [ARMoveSCNViewController new];
            [self presentViewController:vc animated:YES completion:nil];
        }
            break;
        case 30:
        {
            ARRotationSCNViewController *vc = [ARRotationSCNViewController new];
            [self presentViewController:vc animated:YES completion:nil];
        }
            break;
    }
}
@end

File: HelloMacRealityKit/HelloMacRealityKit/ViewController.swift
//
//  ViewController.swift
//  HelloMacRealityKit
//
//  Created by zhoufei on 2023/10/22.
//

import Cocoa
import RealityKit
import RealityFoundation

class ViewController: NSViewController {
    var session: PhotogrammetrySession?

    override func viewDidLoad() {
        super.viewDidLoad()

        // Do any additional setup after loading the view.
        let inputUrl = URL(fileURLWithPath: "/Users/zhoufei/Desktop/FruitCakeSlice/", isDirectory: true)
        session = try! PhotogrammetrySession(input: inputUrl, configuration: PhotogrammetrySession.Configuration())
        
        // Create an async message stream dispatcher task

        Task {
            do {
                for try await output in session!.outputs {
                    switch output {
                    case .requestProgress(let request, let fraction):
                        print("Request progress: \(fraction)")
                    case .requestComplete(let request, let result):
                        if case .modelFile(let url) = result {
                            print("Request result output at \(url).")
                        }
                    case .requestError(let request, let error):
                        print("Error: \(request) error=\(error)")
                    case .processingComplete:
                        print("Completed!")
                        modalFileOutPut()
                    default:  // Or handle other messages...
                        break
                    }
                }
            } catch {
               print("Fatal session error! \(error)")
            }
        }
        
    }
    
    func modalFileOutPut() {
        try! session!.process(requests: [
            .modelFile(url: URL(filePath: "/Users/zhoufei/Desktop/Cake/model-reduced.usdz"), detail: .reduced),
            .modelFile(url: URL(filePath: "/Users/zhoufei/Desktop/Cake/model-reduced.usdz"), detail: .medium)
        ])
    }

    override var representedObject: Any? {
        didSet {
        // Update the view, if already loaded.
        }
    }


}

